plugins/a11y-keyboard/msd-a11y-keyboard-atspi.c

```diff
@@ -0,0 +1,158 @@
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "config.h"
+
+#include "msd-a11y-keyboard-atspi.h"
+
+#include <glib-object.h>
+#include <gdk/gdk.h>
+#include <atspi/atspi.h>
+
+
+struct _MsdA11yKeyboardAtspi
+{
+        GObject              parent;
+        AtspiDeviceListener *listener;
+        gboolean             listening;
+};
+
+G_DEFINE_TYPE (MsdA11yKeyboardAtspi, msd_a11y_keyboard_atspi, G_TYPE_OBJECT)
+
+static void
+msd_a11y_keyboard_atspi_finalize (GObject *obj)
+{
+        MsdA11yKeyboardAtspi *self = MSD_A11Y_KEYBOARD_ATSPI (obj);
+
+        g_clear_object (&self->listener);
+        self->listening = FALSE;
+
+        G_OBJECT_CLASS (msd_a11y_keyboard_atspi_parent_class)->finalize (obj);
+}
+
+static void
+msd_a11y_keyboard_atspi_class_init (MsdA11yKeyboardAtspiClass *klass)
+{
+        GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+        object_class->finalize = msd_a11y_keyboard_atspi_finalize;
+}
+
+static gboolean
+on_key_press_event (const AtspiDeviceEvent *event,
+                    void                   *user_data G_GNUC_UNUSED)
+{
+        /* don't ring on capslock itself, that's taken care of by togglekeys
+         * if the user want it. */
+        if (event->id == GDK_KEY_Caps_Lock)
+                return FALSE;
+
+        gdk_display_beep (gdk_display_get_default ());
+
+        return FALSE;
+}
+
+static void
+msd_a11y_keyboard_atspi_init (MsdA11yKeyboardAtspi *self)
+{
+        self->listener = NULL;
+        self->listening = FALSE;
+
+#ifndef DESTROYING_ATSPI_LISTENER_DOES_NOT_CRASH
+        /* init AT-SPI if needed */
+        atspi_init ();
+
+        self->listener = atspi_device_listener_new (on_key_press_event,
+                                                    self, NULL);
+        /* leak a reference so that this listener is *never* destroyed, to
+         * prevent the crash even if our object gets destroyed.
+         * See https://gitlab.gnome.org/GNOME/at-spi2-core/-/issues/22 */
+        g_object_ref (self->listener);
+#endif
+}
+
+static void
+register_deregister_events (MsdA11yKeyboardAtspi *self,
+                            gboolean              do_register)
+{
+        g_return_if_fail (MSD_IS_A11Y_KEYBOARD_ATSPI (self));
+        g_return_if_fail (ATSPI_IS_DEVICE_LISTENER (self->listener));
+
+        /* register listeners for all keys with CAPS_LOCK modifier */
+        for (AtspiKeyMaskType mod_mask = 0; mod_mask < 256; mod_mask++)
+        {
+                if (! (mod_mask & (1 << ATSPI_MODIFIER_SHIFTLOCK)))
```

Comment:
`(1 << ATSPI_MODIFIER_SHIFTLOCK)` == `SPI_KEYMASK_SHIFTLOCK`?
```
registryd/keymasks.h:#define SPI_KEYMASK_SHIFTLOCK (1<<1)  /* LockMask */
```
If they are the same I prefer `SPI_KEYMASK_SHIFTLOCK`

Reply:
It's not part of the API, is it?

Reply:
No, it's not available in the [API](https://developer.gnome.org/libatspi/stable/libatspi-atspi-constants.html), therefore it's better to use `1 << ATSPI_MODIFIER_SHIFTLOCK` as mask maker, instead of adding `#define SPI_KEYMASK_SHIFTLOCK (1<<1)` a the top of the file, as a maker for (de)registering key event listeners.
```
#define ATSPI_MODIFIERTYPE_COUNT (7+1)
```
2^8=256, possible masks and you only (de)register those that have the bit ATSPI_MODIFIER_SHIFTLOCK set to 1, ie 128 event listeners.

I don’t think it’s the most efficient way to deal with a modifier key mask but it seems valid at first glance.

Reply:
Yeah I know this API seems a bit weird, but it seems to be how it's supposed to be used -- and how other clients like Orca deal with it.